### DISCUSSION TITLE
Shared printer not processing jobs for non-admin users

### DISCUSSION THREAD

**Discussion by lravelo** [clawsoftware/clawPDF#90] [Q&A] [answered]
so I've installed the printer as documented in the wiki and it works if the user sending the print job is an admin. any print jobs sent by regular users errors out on the client machine and don't generate any logs on the print server.  Not seeing any documentation here on any troubleshooting steps so not sure where to look.

**Comment by lravelo** [comment]
Tried this with both a local account on the print server as well as a domain account (neither of which have admin privileges) and the end result is the same.

**Reply by clawsoftware** [reply]
@lravelo Unfortunately, I cannot reproduce this behavior. Have you tried to connect the shared printer manually under the Non Admin User?

**Reply by lravelo** [reply]
> @lravelo Unfortunately, I cannot reproduce this behavior. Have you tried to connect the shared printer manually under the Non Admin User?

yes I have. the printer installs just fine if I install it as a shared printer (i.e., specifying in the  \\server\printer format) but once I go to print, anything sent to it from a non-admin account immediately errors out.  nothing is sent to the print server so there are no logs generated on that side.

the one thing I do see in the event viewer, however, is a DistributedCOM Event ID 10016 warning any time I sent any sort of print job.

**Reply by clawsoftware** [reply]
@lravelo Can you please enable print logging on a client and let me know the possible error output.

https://www.manageengine.com/products/active-directory-audit/how-to/how-to-check-print-server-logs.html
https://www.papercut.com/kb/Main/LogPrintJobsInEventViewer

**Reply by lravelo** [reply]
> @lravelo Can you please enable print logging on a client and let me know the possible error output.
> 
> https://www.manageengine.com/products/active-directory-audit/how-to/how-to-check-print-server-logs.html https://www.papercut.com/kb/Main/LogPrintJobsInEventViewer

so just did another test print and there are no error messages in that printservice operational log.  log basically says that the print job went through successfully. that DCOM 10016 error message still pops up any time I attempt to print to that shared printer.  Let me try switching the printer back to a domain account instead of the local user I created and check to see if the logs show something different.

EDIT: can confirm that the behavior is identical regardless if using a local account on the print server or using a domain account.

**Reply by clawsoftware** [reply]
@lravelo Thx. What server and client version are you using and is the client x86 or x64? I will set up a new test environment.

**Reply by lravelo** [reply]
@clawsoftware the server is Windows Server 2019 Version 1809.  Client machine is a test Windows 10 x64 Build 21H1 virtual machine

**Comment by ft3pierce** [comment]
i am getting the same issue, I am using a windows 2019 server system as network print server. I think it is a permissions issue on the server it self. i have tried giving the non-admin user full permissions on the printer and in the clawpdf directory in the program files (x86) directory, but that made no difference.

**Reply by ft3pierce** [reply]
i get this in the printservice logs "The document Print Document, owned by XXXX, failed to print on printer pdfcreator. Try to print the document again, or restart the print spooler. 
Data type: NT EMF 1.008. Size of the spool file in bytes: 6784340. Number of bytes printed: 0. Total number of pages in the document: 1. Number of pages printed: 0. Client computer: XXXXX. Win32 error code returned by the print processor: 1006. The volume for a file has been externally altered so that the opened file is no longer valid.

**Comment by clawsoftware** [comment]
@ft3pierce @lravelo The client and print server are in the same domain and the printing user is also a domain user in the same domain?

**Reply by lravelo** [reply]
yes same domain.  as mentioned as well I created a local user on the print server itself and that didn't work either.  I'm working on enabling the print logging to get more info.

**Reply by ft3pierce** [reply]
yes, same domain

**Reply by ft3pierce** [reply]
client is x64

**Reply by ft3pierce** [reply]
client is windows 11

**Comment by clawsoftware** [comment]
@ft3pierce @lravelo Ok, thx. My print server is Windows 2022, I will configure a Windows 2019 Server and test it. Have you made any hardening regarding the PrintNightmare vulnerability that is causing the problem? Do other shared printers work on the print server?

**Reply by lravelo** [reply]
> @ft3pierce @lravelo Ok, thx. My print server is Windows 2022, I will configure a Windows 2019 Server and test it. Have you made any hardening regarding the PrintNightmare vulnerability that is causing the problem? Do other shared printers work on the print server?

I do have the point and print restrictions GPO configured for all of my endpoints.
These are the items configured there:

- Users can only point and print to these servers = Enabled
- Enter fully qualified server names seperated by semicolons = (my print server name is listed here)
- Users can only point and print to machines in their forest = Disabled

Security Prompts
- When installing drivers for a new connection = Do not show warning or elevation prompt
- When updating drivers for an existing connection = Show warning and elevation prompt

All my other printers have installed fine with this config

**Reply by ft3pierce** [reply]
other shared printers work fine

**Comment by ft3pierce** [comment]
i have done a test and made it work, this is what i did

non admin user could not print to networked shared klawpdf
until i did this

step 1 - i put the non admin user  into the domain admins group and then logged in to the windows 11 system and printed a test page
             this printed. (as expected)
step 2 - logged out of windows 11 system and then removed the user from domain admins group 
step3 - logged back into window 11 system and printed test page.
            this printed 

hope this helps

**Reply by ft3pierce** [reply]
further to this i logged the same non admin user into a different windows system and connected to the networked klawpdf printer and printed successfully with out having to go thru the previous steps, so it seems that the non-admin only has to do it on one system and then it will work on all  other systems he/she logs into

hope this helps

**Reply by clawsoftware** [reply]
Thanks to you @ft3pierce! A very strange behavior. Your workaround should help me narrow down the problem.

**Reply by lravelo** [reply]
@ft3pierce so what you are saying is you logged into a windows endpoint as the same service account used on the print server, installed the printer under the context of that same user and then sent the print job from that same user as well? what happens when you switch to a different account and then install the printer under that account?

**Reply by ft3pierce** [reply]
did not use the the service account, just added any non-admin user to the domain admins group, logged in and printed once, logged out then remove the that user from the domain admins group, logged back into end point system, using the same user account and it continued to be able to print. this would need to be done for every user wanting to print to the klawpdf printer on the print server. and once a user done this on one end point system it would work for any system that user account logged into.

**Reply by clawsoftware** [reply]
Since I can't reproduce the problem yet, it would be interesting to know if it's enough that the user is temporarily added to the Local Computer Administrators group instead of Domain Administrators.

**Reply by ft3pierce** [reply]
do you local admin group on print server?  my print server is also a domain controller, so it does not have a local admins group, hence why i added to domain admins group. adding to local admins group on end point system had no effect.

**Reply by lravelo** [reply]
> Since I can't reproduce the problem yet, it would be interesting to know if it's enough that the user is temporarily added to the Local Computer Administrators group instead of Domain Administrators.

I just did this test and it still had the same issue.

**Reply by ft3pierce** [reply]
did it work when the user was in the local adminstrators group on the print server?

**Reply by lravelo** [reply]
> did it work when the user was in the local adminstrators group on the print server?

it did not, unfortunately.

**Comment by clawsoftware** [comment]
Guys I need your help @lravelo @ft3pierce. I have now installed a new test environment and everything works as it should. I have done the following steps:

- Clean Windows 11 and Server 2019 install
- Updated both to last patch level
- Installed AD, DNS, DHCP and Print role on Server 2019.
- Joined Windows 11 to the domain
- Installed clawPDF on the Server 2019 with the domain admin
- Created domain user srv_clawpdf and a test user (both domain users)
- Allowed user srv_clawpdf to log in locally on Server 2019 in the GPO "Defualt Domain Controller Policy".
- Stored srv_clawpdf with clawPDF.Bridge.exe
- Print Profile created in clawPDF with srv_clawpdf and assigned to the printer
- Shared the printer with default settings
- Logged on to Windows 11 with the test user and connected the shared printer (driver was installed with authentication of the local Windows 11 admin)
- Printed a test page with the test user without problems

**Reply by lravelo** [reply]
I will test this shortly.  What about in a scenario where the print server is not installed on the domain controller (which is a more likely deployment given that it's best practice to leave additional roles off of them)?

**Reply by clawsoftware** [reply]
Yes, my other test environment is a 2022 server with only the print server role in a 2022 domain, as it is best practice to separate the roles. However, since @ft3pierce has all roles on one server, I wanted to test this scenario as well.

**Reply by lravelo** [reply]
So I've removed everything and started from scratch

- Installed clawpdf 0.92 on my print server using a domain admin account (originally I installed this using the local administrator on the print server)
- I have a domain service account and a test domain account.  Neither have domain admin nor local admin privileges in either the print server or my test client device
- I ran the following to enable the network printing `clawPDF.Bridge.exe /networkprinter=enable /username=clawpdf-svc /domain=domain.local` and entered the credentials
- Created a test profile
- Created a new printer and attached the profile
- Shared the printer via the print management mmc console
- Logged into the test VM with my test user account and installed the printer
- Sent the test print job

Unfortunately the end result is the same.

**Reply by lravelo** [reply]
also tried to deploy via GPO to the machine itself rather than the user and the result is the same.

**Reply by clawsoftware** [reply]
@lravelo @ft3pierce I think I have found the problem. This appears only in some cases. I will get back to you as soon as I have created a patch.

**Reply by ft3pierce** [reply]
ok thanks

**Comment by clawsoftware** [comment, marked as answer]
@lravelo  @ft3pierce Thanks again for your help! It was unfortunately not easy to find the problem, but with the version the issue should hopefully be resolved: https://github.com/clawsoftware/clawPDF/raw/master/src/_MSI/clawPDF_0.9.3_setup.msi

**Reply by ft3pierce** [reply]
I completely uninstalled the ver0.9.2 from the server and installed ver0.9.3 and tried in on a different client system, but unfortunately, I get the same error as before.

**Reply by ft3pierce** [reply]
although, it did continue to work on my original testing client system. it's like the user needs local admin access to the print server during the first time it prints or the user needs local admin access to the print server to download the printer driver files from the print server.

**Reply by hessandrew** [reply]
@ft3pierce This is the default behavior that an admin must install the driver. You can change the behavior with these settings: https://theitbros.com/allow-non-admins-install-printer-drivers-via-gpo/

**Reply by ft3pierce** [reply]
the user I used to test this with is a local admin on the client system.

**Reply by clawsoftware** [reply]
@ft3pierce

Please follow these steps so that I can get a debugging log:

- Set the LogLevel registry key on the print server to 3 under HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Control\Print\Monitors\CLAWMON > LogLevel .
- Restart the print spooler
- Try to print on the client
- Attach the contents of the C:\Windows\System32\clawmon.log file here

**Reply by ft3pierce** [reply]
ok after doing that and then looking at the log file, i could see that clawpdf was creating files and folders in C:\Windows\TEMP\clawPDF\Spool\ , I added  full security permissions to the this folder for my test user and that user can now print. I also tested with a different user and that user worked correctly as well, so it is just a permissions issue with that folder on the server.

**Reply by ft3pierce** [reply]
attached is the C:\Windows\System32\clawmon.log file
[clawmon.log](https://github.com/clawsoftware/clawPDF/files/11488219/clawmon.log)

**Reply by clawsoftware** [reply]
@ft3pierce I have fixed this in version 0.9.3. It seems that the driver from version 0.9.2 is still on the print server. Please uninstall all versions and reinstall version 0.9.3. It could be security problematic to give all rights to the folder. In my patch I manage it differently so that there are no security problems.

**Reply by clawsoftware** [reply]
@ft3pierce Thanks for the test and good to know it is the problem.

**Reply by lravelo** [reply]
@clawsoftware I'm trying to remove the old driver from the print server but no matter which way I try it fails.  Windows keeps complaining about how the printer driver is currently in use despite the fact that I've removed the application from the print server.  I'm having the same issue on the client side as well

Disregard.  Restarted the spooler allowed me to remove it on the print server

**Reply by clawsoftware** [reply]
@ft3pierce @lravelo 

The driver does not need to be removed from the clients. A manual variant on the print server is:

- uninstall clawPDF
- stop the print spooler
- delete or rename the files "C:\Windows\System32\clawmon.dll" and "C:\Windows\System32\clawmonui.dll
- Start the spooler
- Install clawPDF

Normally the installer did that itself. I am surprised that there are problems.

**Reply by lravelo** [reply]
So this fixed it for me for at least one end user and my test environment:

- Fully removed clawPDF from my print server
- Stopped and started the print spooler and removed the clawPDF driver from the print server
- Reinstalled clawPDF 0.9.3
- Recreated the profile and printer
- Shared the printer
- Installed the shared printer on end user's machine and sent a test print
- Success

**Reply by clawsoftware** [reply]
Very good. Sorry for the trouble it has caused.

**Reply by lravelo** [reply]
> Very good. Sorry for the trouble it has caused.

no trouble at all. thank you for creating this product! it has solved a pretty big and long outstanding issue for me.

**Reply by clawsoftware** [reply]
@lravelo Thanks, I'm glad it is useful. For bugs or feature requests feel free to report.
